python_modules/dagster/dagster/_utils/caching_instance_queryer.py

```diff
@@ -54,10 +55,44 @@ def __init__(self, instance: "DagsterInstance"):
         # materialization record for a >= cursor, we don't need to query the instance
         self._no_materializations_after_cursor_cache: Dict[AssetKeyPartitionKey, int] = {}
 
+        self._asset_record_cache: Dict[AssetKey, Optional[AssetRecord]] = {}
+
     @property
     def instance(self) -> "DagsterInstance":
         return self._instance
 
+    def prefetch_for_keys(self, asset_keys: Sequence[AssetKey]):
+        """For performance, batches together queries for selected assets"""
+        asset_records = self.instance.get_asset_records(asset_keys)
+        for asset_record in asset_records:
+            self._asset_record_cache[asset_record.asset_entry.asset_key] = asset_record
+
+        # get the latest N materialization records. in most cases, this will hit a large portion of
+        # the latest materialization records per asset key, although a more sophisticated method
+        # of bucketing queries could be used here to guarantee that we actually get the exact set
+        # we're aiming for here.
+        latest_records = self._instance.get_event_records(
+            EventRecordsFilter(
+                event_type=DagsterEventType.ASSET_MATERIALIZATION,
+            ),
+            limit=len(asset_keys),
+            ascending=False,
+        )
+        for record in latest_records:
+            asset_partition = AssetKeyPartitionKey(
```

Comment:
I think we might need a check downstream to ensure that these partitions are still valid partition keys of the asset

Reply:
we never actually iterate through the cache, so this shouldn't change the functionality